lab_2_tokenize_by_bpe/main.py

```diff
@@ -81,10 +189,36 @@ def decode(
     :param end_of_word_token: an end-of-word token
     :return: decoded sequence
     """
+    if not isinstance(encoded_text, list) \
+            or not all(isinstance(el, int) for el in encoded_text) \
+            or not isinstance(vocabulary, dict) \
+            or not (isinstance(end_of_word_token, str) or end_of_word_token is None):
+        return None
+
+    encoded_dict = {}
+    for index, num in enumerate(encoded_text):
+        if num not in encoded_dict:
+            encoded_dict[num] = []
+        encoded_dict[num].append(index)
+
+    decoded_text = [''] * len(encoded_text)
+
+    for token, identifier in vocabulary.items():
+        if end_of_word_token and end_of_word_token in token and identifier in encoded_dict:
+            space_indexes = encoded_dict[identifier]
+            for space_index in space_indexes:
+                decoded_text[space_index] = token[:-4] + ' '
```

Comment:
Why 4??? I totaly do not understand!

Reply:
Initially I meant the length of end_of_word_token, as it always was  '</s>'
I corrected it to be more logical way